### DISCUSSION TITLE
Algorithm for Find-Similar

### DISCUSSION THREAD

**Discussion by stevengogogo** [stevengogogo/DSA_EmailSearcher#42] [Ideas 解題思路]
# We need ideas!

Context Similarity
------------------
> The context similarity of two mails is defined as **Jaccard index**, where A, B denotes the **token sets** of the two mails.

<img height= 100 src="https://user-images.githubusercontent.com/29009898/119257578-6ecacc80-bbf8-11eb-9bd5-e7b1067d68cd.png">

Problem Statement
-----------------
- `FIND-SIMILAR(MESSAGE-ID, THRESHOLD)`
    Return the mail IDS that shares a similarity greater than the threshold with such mail and sorted in increasing order.

**Comment by stevengogogo** [comment]
Proposal: **Jaccard Similarity algorithm with minihashing**
--------------------------------

`J(A,B) = ∣A ∩ B∣ / ∣A∣ + ∣B∣ - ∣A ∩ B|`

- minihashing

References
------------
1. [**Main pp.82-92**: SIMILARITY-PRESERVING SUMMARIES OF SETS, Chapter 3.3](https://github.com/stevengogogo/DSA_EmailSearcher/files/6529958/ch3.1.pdf)
2. [minihashing and Jaccard similarity. wiki](https://en.wikipedia.org/wiki/MinHash)
3. [**Naive method**: 9.5.1. The Jaccard Similarity algorithm](https://neo4j.com/docs/graph-algorithms/current/labs-algorithms/jaccard/)

**Reply by stevengogogo** [reply]
@jcgs2503 how do you think? if we use this on hashed tokens (we need low rate of spurious hits)

**Reply by stevengogogo** [reply]
Highly recommand this article:
 [**Main pp.82-92**: SIMILARITY-PRESERVING SUMMARIES OF SETS, Chapter 3.3](https://github.com/stevengogogo/DSA_EmailSearcher/files/6529958/ch3.1.pdf)

Highlights
-----------

1. How to deal with super big set
   - minihashing
2. Use random hash to minimic jaccarb similarity

**Reply by jcgs2503** [reply]
I'll take a look, thanks.

**Reply by stevengogogo** [reply]
https://barnasahadotcom.files.wordpress.com/2016/09/similarity.pdf

**Reply by stevengogogo** [reply]
![image](https://user-images.githubusercontent.com/29009898/123438993-ddd67f00-d603-11eb-8de3-a417b5d6d521.png)

**Reply by stevengogogo** [reply]
https://www.youtube.com/watch?v=96WOGPUgMfw

**Reply by stevengogogo** [reply]
https://towardsdatascience.com/understanding-locality-sensitive-hashing-49f6d1f6134

**Reply by stevengogogo** [reply]
https://towardsdatascience.com/understanding-locality-sensitive-hashing-49f6d1f6134

**Reply by stevengogogo** [reply]
Mod method
-------------
https://cs.wmich.edu/~alfuqaha/summer14/cs6530/lectures/SimilarityAnalysis.pdf

**Reply by stevengogogo** [reply]
mini hash 細節
--------------

http://ethen8181.github.io/machine-learning/clustering_old/text_similarity/text_similarity.html

**Comment by stevengogogo** [comment]
Bitset 
------

|[![image](https://user-images.githubusercontent.com/29009898/121485144-7f55b200-ca02-11eb-94e8-a098a51fb4dc.png)](https://owlcation.com/stem/C-bitset-with-example)|![image](https://user-images.githubusercontent.com/29009898/121485245-985e6300-ca02-11eb-9b38-e06cee9e6511.png)|
|---|---|


> From 助教
> Bitset 是 C++ STL 的 template, 用 preprocessing 的數組可以高效的比較 similarity


可以用來處理 subset 的問題

- Bitset 介紹: https://segmentfault.com/a/1190000020958936
- 用法: https://www.itread01.com/content/1550610730.html
- 找子集合: https://ithelp.ithome.com.tw/articles/10231549
- C 語言: https://vimsky.com/zh-tw/article/58.html
